lib/id_partner.rb

```diff
@@ -0,0 +1,278 @@
+
+      # Prepare the payload for the token exchange request
+      payload = {
+        code: decoded_jwt["code"],
+        code_verifier: proofs[:code_verifier],
+        grant_type: "authorization_code",
+        redirect_uri: config[:redirect_uri]
+      }
+
+      # Send the token exchange request
+      uri = URI(endpoints[:token_endpoint])
+      http = Net::HTTP.new(uri.host, uri.port)
+      headers = {
+        "Authorization" => "Basic #{basic_auth_credentials}",
+        "Content-Type" => "application/x-www-form-urlencoded",
+        "Accept" => "application/json"
+      }
+      token_request = Net::HTTP::Post.new(uri.request_uri, headers)
+      token_request.set_form_data(payload)
+      token_response = http.request(token_request)
+      raise "Failed to exchange token: #{token_response.body}" unless token_response.is_a?(Net::HTTPSuccess)
+
+      JSON.parse(token_response.body)
+    end
+
+    def userinfo(access_token)
+      # Prepare the URI and headers for the userinfo request
+      uri = URI(endpoints[:userinfo_endpoint])
+      http = Net::HTTP.new(uri.host, uri.port)
+      headers = {
+        "Authorization" => "Bearer #{access_token}",
+        "Accept" => "application/json"
+      }
+
+      # Send the userinfo request
+      userinfo_request = Net::HTTP::Get.new(uri.request_uri, headers)
+      userinfo_response = http.request(userinfo_request)
+
+      # Check the response and parse the userinfo data
+      raise "Failed to retrieve userinfo: #{userinfo_response.body}" unless userinfo_response.is_a?(Net::HTTPSuccess)
+
+      JSON.parse(userinfo_response.body)
+    end
+
+    private
+
+    def extract_claims(claims_object)
+      return [] unless claims_object.is_a?(Hash)
+
+      userinfo_keys = claims_object[:userinfo]&.keys || []
+      id_token_keys = claims_object[:id_token]&.keys || []
+      (userinfo_keys + id_token_keys).uniq
+    end
+
+    def obtain_well_known_config_endpoints
+      uri = URI("#{config[:provider_url]}/.well-known/openid-configuration")
+      response = Net::HTTP.get(uri)
+      well_known_config = JSON.parse(response)
+      @endpoints = {
+        authorization_endpoint: well_known_config["authorization_endpoint"],
+        token_endpoint: well_known_config["token_endpoint"],
+        userinfo_endpoint: well_known_config["userinfo_endpoint"],
+        pushed_authorization_request_endpoint: well_known_config["pushed_authorization_request_endpoint"],
+        jwks_uri: well_known_config["jwks_uri"]
+      }
+    rescue StandardError => e
+      raise "Failed to fetch well-known config: #{e.message}"
+    end
+
+    def create_request_object(params)
+      extended_params = params.merge({
+                                       iss: config[:client_id],
+                                       aud: config[:provider_url],
+                                       exp: Time.now.to_i + 60,
```

Comment:
this is in seconds same as node: 
https://github.com/idpartner-app/identity-provider-services/blob/7b5ff5e76bcd413712a58b1df66f5cd8fab2a0ce/packages/trust-platform-services/test/fixtures.js#L199